File: problems/Java/126. Word Ladder II.java
class Solution {
    List<List<String>> res = new ArrayList<>();
    Map<String, List<String>> mp = new HashMap<>();

    public List<List<String>> findLadders(String beginWord, String endWord, List<String> wordList) {
        Set<String> d = new HashSet<>(wordList);
        if (!d.contains(endWord)) return res;
        
        Set<String> cur = new HashSet<>();
        cur.add(beginWord);
        boolean fnd = false;

        while (!cur.isEmpty() && !fnd) {
            d.removeAll(cur);
            Set<String> nxt = new HashSet<>();
            
            for (String w : cur) {
                char[] chs = w.toCharArray();
                for (int i = 0; i < chs.length; i++) {
                    char old = chs[i];
                    for (char c = 'a'; c <= 'z'; c++) {
                        chs[i] = c;
                        String nw = new String(chs);
                        if (d.contains(nw)) {
                            if (nw.equals(endWord)) fnd = true;
                            nxt.add(nw);
                            mp.computeIfAbsent(nw, k -> new ArrayList<>()).add(w);
                        }
                    }
                    chs[i] = old;
                }
            }
            cur = nxt;
        }

        if (fnd) {
            List<String> p = new ArrayList<>();
            p.add(endWord);
            dfs(endWord, beginWord, p);
        }
        return res;
    }

    void dfs(String w, String tgt, List<String> p) {
        if (w.equals(tgt)) {
            List<String> tmp = new ArrayList<>(p);
            Collections.reverse(tmp);
            res.add(tmp);
            return;
        }
        if (!mp.containsKey(w)) return;
        for (String par : mp.get(w)) {
            p.add(par);
            dfs(par, tgt, p);
            p.remove(p.size() - 1);
        }
    }
}

File: problems/Python/126. Word Ladder II.py
class Solution:
    def findLadders(self, beginWord: str, endWord: str, wordList: List[str]) -> List[List[str]]:
        d = set(wordList)
        if endWord not in d: return []
        
        cur = {beginWord}
        mp = collections.defaultdict(list)
        
        while cur:
            if endWord in cur: break
            nxt = set()
            d -= cur
            for w in cur:
                for i in range(len(w)):
                    for c in 'abcdefghijklmnopqrstuvwxyz':
                        nw = w[:i] + c + w[i+1:]
                        if nw in d:
                            nxt.add(nw)
                            mp[nw].append(w)
            cur = nxt
            
        res = []
        def dfs(w, p):
            if w == beginWord:
                res.append([beginWord] + p[::-1])
                return
            for par in mp[w]:
                dfs(par, p + [w])
                
        if endWord in mp:
            dfs(endWord, [])
            
        return res